{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining Nucleotide Encodings\n",
    "```\n",
    "pi:ababaian\n",
    "start: 2016 12 20\n",
    "complete : 2017 01 02\n",
    "```\n",
    "## Introduction to IUPAC Characters\n",
    "\n",
    "   In bioinformatics, nucleotide-sequencing information is usually represented using the characters {A, T, U, C, G, N}. Additionally [IUPAC](http://www.bioinformatics.org/sms/iupac.html) defines {R, Y, M, K, S, W, H, B, V, D, X} for all different combinations of {A, T, C, G}\n",
    "    \n",
    "| IUPAC nucleotide code\t| Base  | Represents |\n",
    "|-----------------------| ----- | ---------- |\n",
    "|A                      |   Adenine      | 1 |\n",
    "|T                      |\tThymine      | 1 |\n",
    "|U                      |\tUracil       | 1 |\n",
    "|C                      |   Cytosine     | 1 |\n",
    "|G                      |   Guanine      | 1 |\n",
    "| | | |\n",
    "|S                      |\tG or C       | 2 |\n",
    "|W                      |\tA or T       | 2 |\n",
    "|R                      |\tA or G       | 2 |\n",
    "|Y                      |\tC or T       | 2 |\n",
    "|M                      |\tA or C       | 2 | \n",
    "|K                      |\tG or T       | 2 |\n",
    "| | | |\n",
    "|B                      |\tC or G or T  | 3 |\n",
    "|D                      |\tA or G or T  | 3 |\n",
    "|H                      |\tA or C or T  | 3 |\n",
    "|V                      |\tA or C or G  | 3 |\n",
    "| | | |\n",
    "|N                      |\tany base     | 4 |\n",
    "|X                      |\tunknown base | 0 |\n",
    "|-                      |\tgap          | 0 |\n",
    "\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASCII Nucleotide Encodings\n",
    "\n",
    "### Standard Encoding (ACTG)\n",
    "The standard or ACTG-encoding for this project is the broadly used {A, T, C, G}. One character perfectly represents one single nucletide. For the 16-bit string {ATGAACGT}, 8-bit substrings in standard encoding are {ATGA} {TGAA} {GAAC} {AACG} {ACGT}.\n",
    "    \n",
    "### Non-standard Encodings\n",
    "Alternative encodings of the nucleotides are provided in the IUPAC format as well as using standard encoding. This is to allow for analysis software to parse the genomes without throwing errors.\n",
    "    \n",
    "#### Strong-Weak (SW)\n",
    "    SW-encoding is {S, W} in IUPAC, but will also be {G, T} respectively.\n",
    "    The 16-bit string {ATGAACGT} can only be incompleteley represented in 8-bits by {WWSWWSSW}\n",
    "\n",
    "#### Pyrimidine-Purine (RY)\n",
    "    RY-encoding is {R, Y} in IUPAC, but will also be {C, A} respectively.\n",
    "    The 16-bit string {ATGAACGT} can only be incompleteley represented in 8-bits by {RYRRRYRY}\n",
    "    \n",
    "#### Amin-Keto (MK)\n",
    "    MK-encoding is {M, K} in IUPAC, but will also be {A, G} respectively.\n",
    "    The 16-bit string {ATGAACGT} can only be incompleteley represented in 8-bits by {MKKMMMKK}\n",
    "    \n",
    "### Lossy Encodings\n",
    "Finally, there are lossy B-, D-, H-, V-encodings.\n",
    "    \n",
    "    B-encoding is {B, A} in IUPAC or {T, A} respectively.\n",
    "    V-encoding is {V, T} in IUPAC or {C, T} respectively. \n",
    "    D-encoding is {D, C} in IUPAC or {T, C} respectively. \n",
    "    H-encoding is {H, G} in IUPAC or {C, G} respectively.\n",
    "    \n",
    "Unlike the Non-standard Encodings, two lossy encodings cannot be combined to recapitulate the complete original information of nucleotide string."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binary Encodings\n",
    "The set {A, T, C, G} can be minimally represented as 2-bits of information. Which means in binary this can be written as permutations of two 1's or 0's. The convention  in .2bit file format is:\n",
    "\n",
    "|        |  A  |  T  |  G  |  C  |\n",
    "|--------|-----|-----|-----|-----|\n",
    "| Binary |  10 |  00 |  11 |  01 |\n",
    "\n",
    "\n",
    "An alternative interpreation of this binary encoding is that the first bit represents if the nucleotide is a Purine or Pyrimidine, and the second bit represents if it is a Strong or Weak nucleotide.\n",
    "    \n",
    "In this way any two non-standard encodings (excluding the lossy-encodings) can completely represent the information in standard-encoding string.\n",
    "\n",
    "\n",
    "|        |  A  |  T  |  G  |  C  |\n",
    "|--------|-----|-----|-----|-----|\n",
    "| Binary |  10 |  00 |  11 |  01 |\n",
    "||||||\n",
    "| Purine |  1  |  0  |  1  |  0  |\n",
    "| Strong |  0  |  0  |  1  |  1  |\n",
    "| Amino  |  1  |  0  |  0  |  1  |\n",
    "\n",
    "\n",
    "For the string 16-bit string:\n",
    "        - Standard: { A T G A A C G T }           16 bits\n",
    "        - 2bit:     { 10 00 11 10 10 01 11 00 }   16 bits\n",
    "        - RY:       { 1 0 1 1 1 0 1 0 }            8 bits\n",
    "        - SW:       { 0 0 1 0 0 1 1 0 }            8 bits\n",
    "        - MK:       { 1 0 0 1 1 1 0 0 }            8 bits\n",
    "        \n",
    "So 2-bit encoding is equivalent to RY-SW encoding, but the alternatives of RY-MK or SW-MK are just as feasible. The information is additive.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biological interpretation of Non-standard Encodings\n",
    "\n",
    "### RY\n",
    "The puRines, A and G, are synthesized by a distinct pathways from the pYrimidines, T and C. Purines and pyrimidines have distinct chemical structures and the mutation rate within the same class of nucleotide (transitions) is higher then a change in nucleotide class (transversions).\n",
    "    \n",
    "This means that given two homologous sequences in the genome, they will accumulate transition differences faster then they accumulate transversion differences. That is to say that for homologous sequences, the RY encoding is expected to be more similar for homologous sequences then any other encodings since transition mutations are not represented.\n",
    "    \n",
    "### SW \n",
    "The 'Strong' nucleotides, guanine and cytosine which form three hydrogen bonds when base-paired are distinct from the 'Weak' nucleotides, adenine and thymine which form 2 hydrogen bonds when base-paired. \n",
    "    \n",
    "Blocks of high GC-content or AT-content, called isochores, stain differently by Giemsa and segregate gene-rich and gene-depleted genomics regions. For a long period of time there is a noted differenece in these blocks of GC content, which means that given a string with high GC content, any adjacent characters are more likely to also have a high GC content.\n",
    " \n",
    "### MK \n",
    "Finally, the IUPAC division of 'Amino' A and C and 'Keto' G and T division is based on a difference in chemical structure of the nucleotides. This is probably the least well-defined grouping of nucleotides compared to Strong-Weak and Purine-Pyrimdine axis."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
